File: 124/main.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxPathSum = function (root) {
  if (root == null) return 0
  let res = -99999
  maxPath(root, res)
  return res

  function maxPath (root, ans) {
    if (root == null) return 0
    let l = Math.max(maxPath(root.left, ans), 0)
    let r = Math.max(maxPath(root.right, ans), 0)
    let sum = l + r + root.val
    res = Math.max(res, sum)
    return Math.max(l, r) + root.val
  }
}

File: 124/main.cpp
#include <bits/stdc++.h>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int ans = INT_MIN;
        maxPath(root, ans);
        return ans;
    }
private:
    int maxPath (TreeNode* root, int &ans) {
        if (!root) return 0;
        int l = max(maxPath(root->left, ans), 0);
        int r = max(maxPath(root->right, ans), 0);
        int sum = l + r + root->val;
        ans = max(ans, sum);
        return max(l, r) + root->val;
    }
};

int main()
{
    Solution sol;
    TreeNode *root = new TreeNode(1);
    int res = sol.maxPathSum(root);
    cout << res << endl;
    return 0;
}
